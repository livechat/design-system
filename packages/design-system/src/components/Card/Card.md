### Cards

Cards are used to group related content together to make LiveChat easier to scan and comprehend.

```js
<div style={{ width: "300px" }}>
  <Card title="Title goes here" img="https://via.placeholder.com/100">
    <div>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore
      </p>
      <Divider />
      <div style={{ paddingRight: "10px", display: "inline-block" }}>
        <Button>Action 1</Button>
      </div>
      <Button>Action 2</Button>
    </div>
  </Card>
</div>
```

```js noeditor
<ComponentHtmlMarkup>
  <div style={{ width: "300px" }}>
    <Card title="Title goes here" img="https://via.placeholder.com/100">
      <div>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
          eiusmod tempor incididunt ut labore
        </p>
        <Divider />
        <div style={{ paddingRight: "10px", display: "inline-block" }}>
          <Button>Action 1</Button>
        </div>
        <Button>Action 2</Button>
      </div>
    </Card>
  </div>
</ComponentHtmlMarkup>
```

### Loading

When a card is loading, provide feedback to the user by display a spinner. Once the data has loaded, the spinner disappears from the card component.

```js
<div style={{ width: "300px" }}>
  <Card title="Title goes here" img="https://via.placeholder.com/100">
    <div style={{ height: "200px", display: "flex", justifyContent: "center" }}>
      <Loader size="medium" />
    </div>
  </Card>
</div>
```

```js noeditor
<ComponentHtmlMarkup>
  <div style={{ width: "300px" }}>
    <Card title="Title goes here" img="https://via.placeholder.com/100">
      <div
        style={{ height: "200px", display: "flex", justifyContent: "center" }}
      >
        <Loader size="medium" />
      </div>
    </Card>
  </div>
</ComponentHtmlMarkup>
```

### Content guidelines

#### TITLE

Card title should be:

- Descriptive: Label the type of content grouped in the card
- Concise and scannable:
  - Use simple, clear language that can be read at a glance
  - Keep titles to single sentence
  - Written in sentence case

#### BUTTON

Buttons should be:

- Clear and predictable: Users should be able to anticipate what will happen when they click a button.
