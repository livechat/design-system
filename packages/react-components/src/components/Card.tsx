import * as React from 'react';
import cx from 'classnames';
import { Text } from './Text';
import { Heading } from './Heading';

export interface ICardProps extends React.HTMLAttributes<HTMLDivElement> {
  title: string;
  img?: string;
}

export const Card: React.FC<ICardProps> = ({
  img,
  title,
  children,
  className,
  ...restProps
}) => {
  return (
    <Text
      as="div"
      size="md"
      className={cx('lc-card', className)}
      {...restProps}
    >
      {(img || title) && (
        <Heading as="div" size="sm" className={'lc-card__title'}>
          {img && <img src={img} className={'lc-card__img'} />}
          {title && <div className={'lc-card__text'}>{title}</div>}
        </Heading>
      )}
      {children}
    </Text>
  );
};
