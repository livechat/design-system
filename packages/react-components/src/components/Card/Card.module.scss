$base-class: card;

.#{$base-class} {
  display: flex;
  flex-direction: column;
  gap: 16px;
  align-items: center;
  border: solid 1px var(--border-subtle);
  border-radius: 4px;
  background-color: var(--background);
  padding: 16px;

  &__header {
    display: flex;
    flex-direction: row;
    gap: 12px;
    align-items: center;
    padding: 0;
    width: 100%;
    height: 48px;

    &__image {
      border-radius: 4px;
      background-color: var(--surface-basic-disabled);
      width: 48px;
      height: 48px;
    }

    &__heading {
      display: flex;
      flex-direction: column;
      align-items: flex-start;
      justify-content: center;

      &__title {
        margin: 0;
        color: var(--content-default);
      }

      &__description {
        margin: 0;
        color: var(--content-subtle);
      }
    }

    &__no-image {
      height: 16px;
    }
  }

  &__content,
  &__expanded-content {
    gap: 10px;
    margin: 0;
    width: 100%;
  }

  &__actions {
    display: flex;
    flex-direction: column;
    gap: 16px;
    align-items: flex-start;
    width: 100%;

    &__line {
      background-color: var(--border-subtle);
      width: 100%;
      height: 1px;
    }

    &__buttons {
      display: flex;
      gap: 12px;
      align-items: center;
      width: 100%;

      &-expander {
        margin-left: auto;
      }
    }
  }
}
