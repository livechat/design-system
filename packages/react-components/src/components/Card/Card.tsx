import * as React from 'react';
import cx from 'clsx';
import {
  ChevronUp,
  ChevronDown,
} from '@livechat/design-system-icons/react/material';

import { Text, Heading } from '../Typography';
import { Button, ButtonProps } from '../Button';
import { Icon } from '../Icon';

import styles from './Card.module.scss';

export type CardButtonOptions = Pick<
  ButtonProps,
  'children' | 'kind' | 'onClick'
>;

export interface CardProps extends React.HTMLAttributes<HTMLDivElement> {
  alt?: string;
  buttonsOptions?: CardButtonOptions[];
  description?: string;
  expandableContent?: React.ReactNode;
  src?: string;
  title?: string;
}

const baseClass = 'card';
const headerClass = `${baseClass}__header`;
const headingClass = `${headerClass}__heading`;
const actionsClass = `${baseClass}__actions`;

export const Card: React.FC<CardProps> = ({
  alt,
  buttonsOptions = [],
  children,
  className,
  description,
  expandableContent,
  src,
  title,
  ...divProps
}) => {
  const [isExpanded, setIsExpanded] = React.useState(false);
  const expandIcon = isExpanded ? ChevronUp : ChevronDown;
  const expandButtonText = isExpanded ? 'Hide' : 'Show more';
  const shouldShowActionButtons = buttonsOptions?.length > 0;
  const shouldShowExpandAction = !!expandableContent;
  const shouldShowActions = shouldShowActionButtons || shouldShowExpandAction;
  const isTitleAvailable = title;

  return (
    <div className={cx(styles[baseClass], className)} {...divProps}>
      {isTitleAvailable && (
        <div className={styles[headerClass]}>
          {src && (
            <img
              alt={alt}
              className={styles[`${headerClass}__image`]}
              src={src}
            />
          )}
          <div className={styles[headingClass]}>
            <Heading size="sm" className={styles[`${headingClass}__title`]}>
              {title}
            </Heading>
            <Text size="sm" className={styles[`${headingClass}__description`]}>
              {description}
            </Text>
          </div>
        </div>
      )}
      <Text as="div" size="md" className={styles[`${baseClass}__content`]}>
        {children}
      </Text>
      {isExpanded && (
        <Text
          as="div"
          size="md"
          className={styles[`${baseClass}__expanded-content`]}
        >
          {expandableContent}
        </Text>
      )}
      {shouldShowActions && (
        <div className={styles[actionsClass]}>
          <div className={styles[`${actionsClass}__line`]} />
          <div className={styles[`${actionsClass}__buttons`]}>
            {shouldShowActionButtons &&
              buttonsOptions.map(
                ({ kind, onClick, children: buttonChildren }) => (
                  <Button size="compact" kind={kind} onClick={onClick}>
                    {buttonChildren}
                  </Button>
                )
              )}
            {shouldShowExpandAction && (
              <Button
                className={styles[`${actionsClass}__buttons-expander`]}
                kind="plain"
                iconPosition="right"
                icon={<Icon source={expandIcon} />}
                onClick={() => setIsExpanded(!isExpanded)}
              >
                {expandButtonText}
              </Button>
            )}
          </div>
        </div>
      )}
    </div>
  );
};
