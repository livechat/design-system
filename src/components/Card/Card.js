import * as React from 'react';
import * as PropTypes from 'prop-types';
import getMergedClassNames from '../../utils/getMergedClassNames';
import styles from './style.scss';

const baseClass = 'card';

const Card = props => {
  const { img, title, children, ...restProps } = props;
  const mergedClassNames = getMergedClassNames(styles[baseClass]);

  return (
    <div {...restProps} className={mergedClassNames}>
      <div className={styles[`${baseClass}__title`]}>
        {img && <img src={img} className={styles[`${baseClass}__img`]} />}
        <div className={styles[`${baseClass}__text`]}>{title}</div>
      </div>
      {children}
    </div>
  );
};

Card.propTypes = {
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  children: PropTypes.node.isRequired,
  img: PropTypes.string
};

export default Card;
