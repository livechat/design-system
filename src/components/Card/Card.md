```js
initialState = {
  selectedDate: null
};

const handleDayClick = (day) => {
  setState({
    selectedDate: day
  }, () => {
    console.log(state.selectedDate)
  })
};
<div style={{display: 'flex'}}>
  <Card
    title="Example card"
  >
    <DatePicker onDayClick={handleDayClick} selectedDays={state.selectedDate}/>
  </Card>
</div>
```

```js
  <Card
    title="Example card with icon"
    img="https://via.placeholder.com/100"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris 
  </Card>
```
```js
<div style={{display: 'flex'}}>
  <Card
    title="Example empty card"
    img="https://via.placeholder.com/100"
  />
</div>
```