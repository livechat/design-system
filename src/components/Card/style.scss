@import '../../scss/variables/colors';

$base-class: 'card';

.#{$base-class} {
  background-color: #fff;
  color: $font-primary-color;
  border: solid 1px $divider-color;
  border-radius: 8px;
  margin-top: 8px;
  padding: 18px;
  padding-bottom: 10px;
  font-size: 15px;
  line-height: 16px;
  overflow: hidden;

  &__title {
    display: flex;
    align-items: center;
    margin-bottom: 14px;
  }

  &__text {
    font-size: 18px;
    line-height: 24px;
    color: $ui-tooltip-color;
    font-weight: 600;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    width: 100%;
  }

  &__img {
    height: 24px;
    width: 24px;
    margin-right: 10px;
    border-radius: 4px;
    display: inline-block;
  }
};
