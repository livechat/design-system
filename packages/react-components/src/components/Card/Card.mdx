import { Canvas, ArgsTable, Meta, Title } from '@storybook/blocks';
import { Figma } from "@storybook/addon-designs/blocks";
import * as CardStories from './Card.stories';
import { BlockColumnsContainer, BlockColumns } from '../../stories/components/BlockColumns';

<Meta of={CardStories} />

<Title>Card</Title>

[Intro](#Intro) | [Component API](#ComponentAPI) | [Content Spec](#ContentSpec)

## Intro <a id="Intro" />

<Canvas of={CardStories.Default} sourceState="none"/>

#### Example implementation
```jsx
<Card
  title="Card title"
  description="Card description"
>
  Card text
</Card>
```

## Component API <a id="ComponentAPI" />
<ArgsTable />

## Content Spec <a id="ContentSpec" />
<Figma showLink height="100%" url="https://www.figma.com/file/9FDwjR8lYvincseDkKypC4/Component-Documentations?type=design&node-id=749-56931&mode=dev" />
