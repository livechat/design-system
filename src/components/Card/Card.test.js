import * as React from 'react';
import { mount } from 'enzyme';
import Card from './Card';

describe('Card', () => {
  it('should render selected Card', () => {
    const component = mount(
      <Card isSelected description="1">
        Agents
      </Card>
    );

    expect(component).toMatchSnapshot();
  });
});
