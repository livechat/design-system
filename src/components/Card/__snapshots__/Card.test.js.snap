// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render Card 1`] = `
<Card
  img="https://via.placeholder.com/100"
  title="Title goes here"
>
  <div
    className="card"
  >
    <div
      className="card__title"
    >
      <img
        className="card__img"
        src="https://via.placeholder.com/100"
      />
      <div
        className="card__text"
      >
        Title goes here
      </div>
    </div>
    test
  </div>
</Card>
`;
